src/topology/uniform_space/UniformSpace.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Reid Barton, Patrick Massot, Scott Morrison
+-/
+
+import category_theory.concrete_category
+import category_theory.full_subcategory
+import category_theory.monad.limits
+import topology.uniform_space.completion
+import topology.Top.basic
+
+/-!
+# The category of uniform spaces
+
+We construct the category of uniform spaces, show that the complete separated uniform spaces
+form a reflective subcategory, and hence possess all limits that uniform spaces do.
+
+TODO: show that uniform spaces actually have all limits!
+-/
+
+
+universes u
+
+open category_theory
+
+/-- A (bundled) uniform spaces. -/
+@[reducible] def UniformSpace : Type (u+1) := bundled uniform_space
+
+namespace UniformSpace
+
+instance (x : UniformSpace) : uniform_space x := x.str
+
+/-- The category instance on `UniformSpace`. -/
+instance concrete_category_uniform_continuous : concrete_category @uniform_continuous :=
+⟨@uniform_continuous_id, by { introsI, apply uniform_continuous.comp; assumption }⟩
+
+def of (α : Type u) [uniform_space α] : UniformSpace := ⟨α⟩
+
+abbreviation forget : UniformSpace.{u} ⥤ Type u := forget
+
+instance forget.faithful : faithful (forget) := {}
+
+/-- The forgetful functor from uniform spaces to topological spaces. -/
+def forget_to_Top : UniformSpace.{u} ⥤ Top.{u} :=
+{ obj := λ X, { α := X.1 },
+  map := λ X Y f, ⟨ f, uniform_continuous.continuous f.property ⟩ }
+
+/-- The natural isomorphism between forgetting from `UniformSpace` to `Type` via `Top`,
+and forgetting directly. -/
+def forget_to_Type_via_Top : forget_to_Top ⋙ Top.forget ≅ forget := iso.refl _
```

Comment:
I wonder if we should just have a generic `forget_comp` for these concrete categories.